nexus/src/app/sagas/snapshot_delete.rs

```diff
@@ -0,0 +1,113 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use super::{ActionRegistry, NexusActionContext, NexusSaga};
+use crate::app::sagas;
+use crate::app::sagas::declare_saga_actions;
+use crate::context::OpContext;
+use crate::{authn, authz, db};
+use serde::Deserialize;
+use serde::Serialize;
+use steno::ActionError;
+use steno::Node;
+
+#[derive(Debug, Deserialize, Serialize)]
+pub struct Params {
+    pub serialized_authn: authn::saga::Serialized,
+    pub authz_snapshot: authz::Snapshot,
+    pub snapshot: db::model::Snapshot,
+}
+
+declare_saga_actions! {
+    snapshot_delete;
+    DELETE_SNAPSHOT_RECORD -> "no_result" {
+        + ssd_delete_snapshot_record
+    }
+}
+
+#[derive(Debug)]
+pub struct SagaSnapshotDelete;
+impl NexusSaga for SagaSnapshotDelete {
+    const NAME: &'static str = "snapshot-delete";
+    type Params = Params;
+
+    fn register_actions(registry: &mut ActionRegistry) {
+        snapshot_delete_register_actions(registry);
+    }
+
+    fn make_saga_dag(
+        params: &Self::Params,
+        mut builder: steno::DagBuilder,
+    ) -> Result<steno::Dag, super::SagaInitError> {
+        builder.append(delete_snapshot_record_action());
+
+        const DELETE_VOLUME_PARAMS: &'static str = "delete_volume_params";
+        const DELETE_VOLUME_DESTINATION_PARAMS: &'static str =
+            "delete_volume_params";
+
+        let volume_delete_params = sagas::volume_delete::Params {
+            volume_id: params.snapshot.volume_id,
+        };
+        builder.append(Node::constant(
+            DELETE_VOLUME_PARAMS,
+            serde_json::to_value(&volume_delete_params).map_err(|e| {
+                super::SagaInitError::SerializeError(
+                    String::from("volume_id"),
+                    e,
+                )
+            })?,
+        ));
+
+        let volume_delete_params = sagas::volume_delete::Params {
+            volume_id: params.snapshot.destination_volume_id.unwrap(),
```

Comment:
Note: Should not be unwrapping, if this can be non-optional

Reply:
I think after PR #2092 it won't be.

Reply:
Removed in https://github.com/oxidecomputer/omicron/pull/2090/commits/3f37a48aa32ef27d1531b832637850078320b503